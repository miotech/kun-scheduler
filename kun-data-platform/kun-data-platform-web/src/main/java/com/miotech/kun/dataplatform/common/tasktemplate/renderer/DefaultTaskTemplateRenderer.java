package com.miotech.kun.dataplatform.common.tasktemplate.renderer;

import com.miotech.kun.dataplatform.model.taskdefinition.TaskConfig;
import com.miotech.kun.dataplatform.model.taskdefinition.TaskDefinition;
import com.miotech.kun.dataplatform.model.tasktemplate.TaskTemplate;
import org.springframework.stereotype.Component;

import java.util.Map;

@Component
public class DefaultTaskTemplateRenderer extends TaskTemplateRenderer {

    @Override
    public TaskConfig render(Map<String, Object> taskConfig, TaskTemplate taskTemplate, TaskDefinition taskDefinition) {
        Map<String, Object> configMap = buildTaskConfig(taskConfig, taskTemplate, taskDefinition);

        return TaskConfig.newBuilder()
                .withParams(configMap)
                .build();
    }
}
