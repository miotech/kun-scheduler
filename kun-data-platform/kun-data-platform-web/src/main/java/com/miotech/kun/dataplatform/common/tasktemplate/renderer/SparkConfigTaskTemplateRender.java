package com.miotech.kun.dataplatform.common.tasktemplate.renderer;

import com.miotech.kun.commons.utils.IdGenerator;
import com.miotech.kun.dataplatform.model.taskdefinition.TaskConfig;
import com.miotech.kun.dataplatform.model.taskdefinition.TaskDefinition;
import com.miotech.kun.dataplatform.model.tasktemplate.ParameterDefinition;
import com.miotech.kun.dataplatform.model.tasktemplate.TaskTemplate;
import com.miotech.kun.workflow.client.model.ConfigKey;
import com.miotech.kun.workflow.operator.SparkConfiguration;
import com.miotech.kun.workflow.utils.JSONUtils;
import org.springframework.stereotype.Component;

import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@Component
public class SparkConfigTaskTemplateRender extends TaskTemplateRenderer {

    private static final String KUN_DATA_PLATFORM_CONFIG_PREFIX = "kun.dataplatform";
    private static final String KUN_DATA_PLATFORM_DATASOURCE_PREFIX = "kun.dataplatform.datasource";
    private static final String SPARK_CONFIG_KEY = "sparkConf";
    private static final String SPARK_JAVA_OPTIONS_KEY = "spark.driver.extraJavaOptions";

    @Override
    public TaskConfig render(Map<String, Object> taskConfig, TaskTemplate taskTemplate, TaskDefinition taskDefinition) {
        Map<String, Object> configMap = buildTaskConfig(taskConfig, taskTemplate, taskDefinition);

        return  TaskConfig.newBuilder()
                .withParams(configMap)
                .build();
    }

    @Override
    public Map<String, Object> buildTaskConfig(Map<String, Object> taskConfig, TaskTemplate taskTemplate, TaskDefinition taskDefinition) {
        Map<String, Object> configMap = super.buildTaskConfig(taskConfig, taskTemplate, taskDefinition);
        for (String key: taskConfig.keySet()) {
            if (!configMap.containsKey(key)) {
                configMap.put(key, taskConfig.get(key));
            }
        }
        List<String> paramKeys = taskTemplate
                .getOperator()
                .getConfigDef()
                .stream()
                .map(ConfigKey::getName)
                .collect(Collectors.toList());

        Map<String, ParameterDefinition> displayParams = taskTemplate
                .getDisplayParameters()
                .stream()
                .collect(Collectors.toMap(ParameterDefinition::getName, Function.identity()));
        Map<String, Object> config = new HashMap<>(configMap);
        if (configMap.containsKey(SPARK_CONFIG_KEY) &&
                paramKeys.contains(SPARK_CONFIG_KEY)) {
            Map<String, String> sparkConfig;
            if (configMap.get(SPARK_CONFIG_KEY) instanceof String) {
                sparkConfig = JSONUtils.jsonStringToStringMap((String) configMap.get(SPARK_CONFIG_KEY));
            } else {
                sparkConfig = (Map<String, String>) configMap.get(SPARK_CONFIG_KEY);
            }

            // add extra parameter into sparkConf as extraJavaOptions
            StringBuilder extraJavaOptions = new StringBuilder(sparkConfig.getOrDefault(SPARK_JAVA_OPTIONS_KEY, ""));

            Map<String, Object> extraConfig = new HashMap<>();
            for (String key: configMap.keySet()) {
                if (!paramKeys.contains(key)) {
                    extraConfig.put(key, configMap.get(key));
                }
                ParameterDefinition parameterDefinition = displayParams.get(key);
                // add datasource as variable
                if (parameterDefinition != null
                        && parameterDefinition.getType().equals("datasource")) {
                    extraJavaOptions.append(String.format(" -D%s.%s=${dataplatform.datasource.%s}", KUN_DATA_PLATFORM_DATASOURCE_PREFIX, configMap.get(key), configMap.get(key)));
                }
            }
            if (!extraConfig.isEmpty()) {
                String encodedString = Base64.getEncoder().encodeToString(JSONUtils.toJsonString(extraConfig).getBytes());
                extraJavaOptions.append(String.format(" -D%s=%s", KUN_DATA_PLATFORM_CONFIG_PREFIX, encodedString));
            }

            sparkConfig.put(SPARK_JAVA_OPTIONS_KEY, extraJavaOptions.toString());
            config.put(SPARK_CONFIG_KEY, JSONUtils.toJsonString(sparkConfig));
        } else {
            config.putAll(configMap);
        }
        config.put(SparkConfiguration.CONF_LIVY_BATCH_NAME, taskDefinition.getName() + " - " + IdGenerator.getInstance().nextId());
        return config;
    }
}
