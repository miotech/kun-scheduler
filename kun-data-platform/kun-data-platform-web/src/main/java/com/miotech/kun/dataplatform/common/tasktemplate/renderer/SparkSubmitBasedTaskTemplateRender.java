package com.miotech.kun.dataplatform.common.tasktemplate.renderer;

import com.google.common.base.Strings;
import com.miotech.kun.dataplatform.model.taskdefinition.TaskConfig;
import com.miotech.kun.dataplatform.model.taskdefinition.TaskDefinition;
import com.miotech.kun.dataplatform.model.tasktemplate.TaskTemplate;
import com.miotech.kun.workflow.client.model.ConfigKey;
import com.miotech.kun.workflow.utils.JSONUtils;

import java.util.*;
import java.util.stream.Collectors;

import static com.miotech.kun.workflow.operator.SparkConfiguration.*;

abstract public class SparkSubmitBasedTaskTemplateRender extends TaskTemplateRenderer {
    protected static final String SPARK_CONFIG_KEY = "sparkConf";
    protected static final String SPARK_MASTER = "master";
    protected static final String SPARK_DEPLOY_MODE = "deployMode";
    protected static final String SPARK_YARN_HOST = "yarnHost";
    protected static final String[] SPARK_SUBMIT_OPTIONS = {"master", "deploy-mode", "proxy-user", "class"};

    @Override
    public TaskConfig render(Map<String, Object> taskConfig, TaskTemplate taskTemplate, TaskDefinition taskDefinition) {
        Map<String, Object> mergedConfig = mergeDefaultConfAndUserDefinedConf(taskConfig, taskTemplate);
        String sparkSubmitParams = buildSparkSubmitCmd(mergedConfig, taskTemplate, taskDefinition);
        Map<String, Object> configMap = new HashMap<>();
        configMap.put(SPARK_BASE_COMMAND, getSparkBaseCmd());
        configMap.put(SPARK_SUBMIT_PARMAS, sparkSubmitParams);

        String master = (String) mergedConfig.get(SPARK_MASTER);
        if (!Strings.isNullOrEmpty(master)) {
            configMap.put(SPARK_MASTER, master);
        }

        String deployMode = (String) mergedConfig.get("deploy-mode");
        if (!Strings.isNullOrEmpty(deployMode)) {
            configMap.put(SPARK_DEPLOY_MODE, deployMode);
        }

        String yarnHost = (String) mergedConfig.get(SPARK_YARN_HOST);
        if (!Strings.isNullOrEmpty(yarnHost)) {
            configMap.put(SPARK_YARN_HOST, yarnHost);
        }

        //s3 credential
        configMap.put(VAR_S3_ACCESS_KEY, CONF_S3_ACCESS_KEY_VALUE_DEFAULT);
        configMap.put(VAR_S3_SECRET_KEY, CONF_S3_SECRET_KEY_VALUE_DEFAULT);

        return TaskConfig.newBuilder()
                .withParams(configMap)
                .build();
    }

    List<String> buildBasicSparkCmd(Map<String, Object> taskConfig, TaskDefinition taskDefinition) {
        List<String> params = new ArrayList<>();

        // parse spark-submit options
        for (String key : SPARK_SUBMIT_OPTIONS) {
            String value = (String) taskConfig.get(key);
            if (!Strings.isNullOrEmpty(value)) {
                params.add("--" + key + " " + value);
            }
        }

        // set spark app name
        params.add("--name" + " " + taskDefinition.getName());

        //parse spark conf
        Map<String, String> sparkConfig;
        if (taskConfig.get(SPARK_CONFIG_KEY) instanceof String) {
            sparkConfig = JSONUtils.jsonStringToStringMap((String) taskConfig.get(SPARK_CONFIG_KEY));
        } else {
            sparkConfig = (Map<String, String>) taskConfig.get(SPARK_CONFIG_KEY);
        }

        if (sparkConfig == null) {
            sparkConfig = new HashMap<>();
        }
        // lineage conf
        String jars;
        if (sparkConfig.containsKey("spark.jars")) {
            jars = sparkConfig.get("spark.jars") + "," + CONF_LINEAGE_JAR_PATH_VALUE_DEFAULT;
        } else {
            jars = CONF_LINEAGE_JAR_PATH_VALUE_DEFAULT;
        }

        //TODO: if lineage jar not configed, adding "spark.sql.queryExecutionListeners" in sparkConf will throw exception
        //TODO: Solution: add this config to operator level
        sparkConfig.put("spark.jars", jars);
        sparkConfig.put("spark.sql.queryExecutionListeners", "za.co.absa.spline.harvester.listener.SplineQueryExecutionListener");
        sparkConfig.put("spark.hadoop.spline.hdfs_dispatcher.address", CONF_LINEAGE_OUTPUT_PATH_VALUE_DEFAULT);
        sparkConfig.put("spark.fs.s3a.access.key", CONF_S3_ACCESS_KEY_VALUE_DEFAULT);
        sparkConfig.put("spark.fs.s3a.secret.key", CONF_S3_SECRET_KEY_VALUE_DEFAULT);

        for (Map.Entry<String, String> entry : sparkConfig.entrySet()) {
            params.add("--conf " + entry.getKey() + "=" + entry.getValue());
        }

        return params;
    }

    public Map<String, Object> mergeDefaultConfAndUserDefinedConf(Map<String, Object> taskConfig, TaskTemplate taskTemplate) {
        Map<String, Object> configMap = new HashMap<>();

        Map<String, Object> defaultValues = taskTemplate.getDefaultValues();
        if (taskConfig == null) {
            configMap.putAll(defaultValues);
        } else {
            Set<String> paramKeys = taskTemplate
                    .getOperator()
                    .getConfigDef()
                    .stream()
                    .map(ConfigKey::getName)
                    .collect(Collectors.toSet());
            paramKeys.addAll(taskConfig.keySet());

            for (String key : paramKeys) {
                Object mergedValue = mergeDefaultAndUserDefinedConfig(defaultValues, taskConfig, key);
                if (mergedValue != null)
                    configMap.put(key, mergedValue);
            }
        }
        return configMap;
    }

    public abstract String buildSparkSubmitCmd(Map<String, Object> taskConfig, TaskTemplate taskTemplate, TaskDefinition taskDefinition);
    public abstract String getSparkBaseCmd();

}
