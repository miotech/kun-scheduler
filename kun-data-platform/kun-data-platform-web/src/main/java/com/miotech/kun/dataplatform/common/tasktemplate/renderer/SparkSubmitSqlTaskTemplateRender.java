package com.miotech.kun.dataplatform.common.tasktemplate.renderer;

import com.miotech.kun.dataplatform.model.taskdefinition.TaskDefinition;
import com.miotech.kun.dataplatform.model.tasktemplate.TaskTemplate;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;

@Component
public class SparkSubmitSqlTaskTemplateRender extends SparkSubmitBasedTaskTemplateRender {

    @Override
    public String buildSparkSubmitCmd(Map<String, Object> taskConfig, TaskTemplate taskTemplate, TaskDefinition taskDefinition) {
        List<String> params = buildBasicSparkCmd(taskConfig);
        params.add(0, "spark-sql");

        //parse sql
        String sql = (String) taskConfig.get("sparkSQL");
        params.add("-e " + "'" + sql + "'");
        return String.join(" ", params);
    }

}
