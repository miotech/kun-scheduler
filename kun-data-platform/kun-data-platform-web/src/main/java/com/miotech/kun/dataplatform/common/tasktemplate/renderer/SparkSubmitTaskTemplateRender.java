package com.miotech.kun.dataplatform.common.tasktemplate.renderer;

import com.miotech.kun.dataplatform.model.taskdefinition.TaskDefinition;
import com.miotech.kun.dataplatform.model.tasktemplate.TaskTemplate;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;

@Component
public class SparkSubmitTaskTemplateRender extends SparkSubmitBasedTaskTemplateRender {

    @Override
    public String buildSparkSubmitCmd(Map<String, Object> taskConfig, TaskTemplate taskTemplate, TaskDefinition taskDefinition) {
        List<String> params = buildBasicSparkCmd(taskConfig, taskDefinition);
        params.add(0, "spark-submit");

        //parse application main and args
        String appMain = (String) taskConfig.get("application");
        String appArgs = (String) taskConfig.get("args");
        params.add(appMain);
        if(appArgs != null){
            params.add(appArgs);
        }
        return String.join(" ", params);
    }

}
