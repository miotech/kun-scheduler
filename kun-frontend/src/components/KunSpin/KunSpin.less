.kun-spin {
  .max-size();
}

#root .load-container {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  top: 0;
  left: 0;
  margin: 0;

  .container{
    width: 50px;
    height: 60px;
    text-align: center;
    font-size: 10px;
    .boxLoading {
      background-color: @darkBlue;
      height: 100%;
      width: 6px;
      display: inline-block;

      -webkit-animation: stretchdelay 1.2s infinite ease-in-out;
      animation: stretchdelay 1.2s infinite ease-in-out;
    }
    .boxLoading2 {
      -webkit-animation-delay: -1.1s;
      animation-delay: -1.1s;
    }
    .boxLoading3 {
      -webkit-animation-delay: -1.0s;
      animation-delay: -1.0s;
    }
    .boxLoading4 {
      -webkit-animation-delay: -0.9s;
      animation-delay: -0.9s;
    }
    .boxLoading5 {
      -webkit-animation-delay: -0.8s;
      animation-delay: -0.8s;
    }
  }
}

.boxLoading,.boxLoading2,.boxLoading3,.boxLoading4,.boxLoading5 {
  margin: 0 2px;
}

#root .kun-spin .ant-spin-text {
  top: calc(50% + 36px);
}

#root .kun-spin-wrapper {
  position: static;
}

#root .kun-spin-wrapper.kun-spin-wrapper--as-block {
  position: relative;
}

#root .load-container--as-block {
  position: relative;
  width: 100%;
  height: 250px;

  .ant-spin-text {
    top: calc(50% + 36px);
    position: absolute;
    left: 50%;
    transform: translate(-50%,0);
  }
}

@-webkit-keyframes stretchdelay {
  0%, 40%, 100% { -webkit-transform: scaleY(0.4) }
  20% { -webkit-transform: scaleY(1.0) }
}

@keyframes stretchdelay {
  0%, 40%, 100% {
    transform: scaleY(0.4);
    -webkit-transform: scaleY(0.4);
  }  20% {
       transform: scaleY(1.0);
       -webkit-transform: scaleY(1.0);
     }
}
