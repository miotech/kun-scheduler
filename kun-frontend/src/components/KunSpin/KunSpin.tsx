import React from 'react';
import { Spin } from 'antd';
import c from 'classnames';
import { SpinProps } from 'antd/es/spin';

import './KunSpin.less';

interface KunSpinProps extends SpinProps {
  asBlock?: boolean;
}

const defaultIndicator = (
  <div className="load-container">
    <div className="container">
      <div className="boxLoading boxLoading1" />
      <div className="boxLoading boxLoading2" />
      <div className="boxLoading boxLoading3" />
      <div className="boxLoading boxLoading4" />
      <div className="boxLoading boxLoading5" />
    </div>
  </div>
);

export const KunSpin: React.FC<KunSpinProps> = props => {
  const { className, wrapperClassName, asBlock, ...restProps } = props;

  return (
    <Spin
      wrapperClassName={c('kun-spin-wrapper', wrapperClassName, {
        'kun-spin-wrapper--as-block': !!asBlock,
      })}
      className={c('kun-spin', className, {
        'load-container--as-block': !!asBlock,
      })}
      indicator={defaultIndicator}
      {...restProps}
    />
  );
};
