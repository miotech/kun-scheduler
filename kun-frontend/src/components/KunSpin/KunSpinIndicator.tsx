import React from 'react';

import './KunSpin.less';

export const KunSpinIndicator = (
  <div className="load-container">
    <div className="container">
      <div className="boxLoading boxLoading1" />
      <div className="boxLoading boxLoading2" />
      <div className="boxLoading boxLoading3" />
      <div className="boxLoading boxLoading4" />
      <div className="boxLoading boxLoading5" />
    </div>
  </div>
);
