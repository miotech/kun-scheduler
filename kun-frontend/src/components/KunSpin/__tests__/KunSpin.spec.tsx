import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { KunSpin } from '../KunSpin';

test('renders correctly',  () => {
  const component = renderer.create(<KunSpin />).toJSON();
  expect(component).toMatchSnapshot();
});

test('mounts correctly', async () => {
  expect(() => {
    mount(<KunSpin />);
  }).not.toThrow();
});
