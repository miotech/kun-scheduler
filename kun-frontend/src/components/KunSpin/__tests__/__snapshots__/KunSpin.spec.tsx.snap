// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="ant-spin ant-spin-spinning kun-spin"
>
  <div
    className="load-container ant-spin-dot"
  >
    <div
      className="container"
    >
      <div
        className="boxLoading boxLoading1"
      />
      <div
        className="boxLoading boxLoading2"
      />
      <div
        className="boxLoading boxLoading3"
      />
      <div
        className="boxLoading boxLoading4"
      />
      <div
        className="boxLoading boxLoading5"
      />
    </div>
  </div>
</div>
`;
